apps/build-notification/src/githubPrComment.ts

```diff
@@ -0,0 +1,133 @@
+import type { Octokit } from "@octokit/rest";
+
+import { Build, type PullRequest } from "@argos-ci/database/models";
+import { getRedisLock } from "@argos-ci/web";
+
+const capitalize = (str: string) => str.charAt(0).toUpperCase() + str.slice(1);
+
+export const formatStatus = (status: string | null | undefined) => {
+  switch (status) {
+    case "stable":
+      return "✅ Stable";
+
+    case "accepted":
+      return "👍 Approved";
+
+    case "error":
+    case "expired":
+    case "rejected":
+    case "aborted":
+      return `❌ ${capitalize(status)}`;
+
+    case "pending":
+    case "progress":
+      return "⏳ Pending";
+
+    case "diffDetected":
+      return "⚠️ Review required";
+
+    default:
+      return `❔ ${status ? capitalize(status) : "Unknown"}`;
+  }
+};
+export const getAggregateStatuses = async (builds: Build[]) => {
+  const buildIds = builds.map((build) => build.id);
+  const statuses = await Build.getStatuses(builds);
+  const conclusions = await Build.getConclusions(buildIds, statuses);
+  const reviewStatuses = await Build.getReviewStatuses(buildIds, conclusions);
+  return builds.map((_build, index) => {
+    if (reviewStatuses[index]) return reviewStatuses[index];
+    if (conclusions[index]) return conclusions[index];
+    return statuses[index];
+  });
+};
+
+const createPrComment = async ({
+  body,
+  number,
+  octokit,
+  owner,
+  repo,
+}: {
+  body: string;
+  number: number;
+  octokit: Octokit;
+  owner: string;
+  repo: string;
+}) => {
+  return await octokit.issues.createComment({
+    owner,
+    repo,
+    issue_number: number,
+    body,
+  });
+};
+
+const updatePrComment = async ({
+  body,
+  commentId,
+  octokit,
+  owner,
+  repo,
+}: {
+  body: string;
+  commentId: number;
+  octokit: Octokit;
+  owner: string;
+  repo: string;
+}) => {
+  return octokit.issues.updateComment({
+    owner,
+    repo,
+    comment_id: commentId,
+    body,
+  });
+};
+
+export const createOrUpdatePrComment = async ({
+  githubAccountLogin,
+  octokit,
+  pullRequest,
+  repositoryName,
+  message,
+}: {
+  commit: string;
+  githubAccountLogin: string;
+  octokit: Octokit;
+  pullRequest: PullRequest;
+  repositoryName: string;
+  message: string;
+}) => {
+  try {
+    const { commentId, number } = pullRequest;
+    const lock = await getRedisLock();
+    await lock.acquire(pullRequest.id, async () => {
```

Comment:
Not the good place to aquire lock. If it blocks at this line in two jobs, two PR comments will be created. You have to do in the lock:

- Try to get existing commentId
- Create comment

The update should not be part of the lock. The lock is to get a comment id.